# Regressão 01 - tarefa 02 - transformações em X e Y

### 1. Melhorando o ajuste no percentual de gorjetas

a. Vamos partir do modelo feito na última tarefa, o que relaciona ```tip_pct``` e ```net_bill```. Carregue a base, os pacotes e reproduza este modelo aqui.

In [2]:
# Carregando os pacotes necessários
import pandas as pd
import seaborn as sns
import statsmodels.formula.api as smf
import numpy as np

# Carregando a base de dados
tips = sns.load_dataset("tips")
tips['tip_pct'] = tips['tip'] / (tips['total_bill'] - tips['tip'])
tips['net_bill'] = tips['total_bill'] - tips['tip']

# Ajustando o modelo de regressão
reg = smf.ols('tip_pct ~ net_bill', data=tips).fit()
print(reg.summary())

                            OLS Regression Results                            
Dep. Variable:                tip_pct   R-squared:                       0.091
Model:                            OLS   Adj. R-squared:                  0.088
Method:                 Least Squares   F-statistic:                     24.31
Date:                Wed, 06 Sep 2023   Prob (F-statistic):           1.52e-06
Time:                        17:50:27   Log-Likelihood:                 108.00
No. Observations:                 244   AIC:                            -212.0
Df Residuals:                     242   BIC:                            -205.0
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      0.3053      0.023     13.166      0.0

### 2. ajuste o modelo de ```tip_pct``` em função do logaritmo de ```net_bill```.

In [3]:
# Ajustando o modelo de regressão com o logaritmo de net_bill
reg_log = smf.ols('tip_pct ~ np.log(net_bill)', data=tips).fit()
print(reg_log.summary())

                            OLS Regression Results                            
Dep. Variable:                tip_pct   R-squared:                       0.202
Model:                            OLS   Adj. R-squared:                  0.198
Method:                 Least Squares   F-statistic:                     61.16
Date:                Wed, 06 Sep 2023   Prob (F-statistic):           1.62e-13
Time:                        17:50:35   Log-Likelihood:                 123.81
No. Observations:                 244   AIC:                            -243.6
Df Residuals:                     242   BIC:                            -236.6
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                       coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------
Intercept            0.6181      0.054  

###  3. ajuste o modelo de ```tip_pct``` em função de um polinômio de segundo grau em ```net_bill```.

In [4]:
# Ajustando o modelo de regressão com polinômio de segundo grau de net_bill
reg_poly = smf.ols('tip_pct ~ net_bill + np.power(net_bill, 2)', data=tips).fit()
print(reg_poly.summary())

                            OLS Regression Results                            
Dep. Variable:                tip_pct   R-squared:                       0.158
Model:                            OLS   Adj. R-squared:                  0.151
Method:                 Least Squares   F-statistic:                     22.55
Date:                Wed, 06 Sep 2023   Prob (F-statistic):           1.06e-09
Time:                        17:51:06   Log-Likelihood:                 117.25
No. Observations:                 244   AIC:                            -228.5
Df Residuals:                     241   BIC:                            -218.0
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                            coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------------
Intercept                 0.48

###  4. ajuste um modelo no log do ```tip_pct``` em função de ```net_bill```. Calcule o $R^2$ também em ```tip_pct```

In [5]:
# Ajustando o modelo de regressão usando log de tip_pct em função de net_bill
tips = tips[tips['tip_pct'] > 0]  # Removendo qualquer valor 0 para evitar log(0)
reg_log = smf.ols('np.log(tip_pct) ~ net_bill', data=tips).fit()
print(reg_log.summary())

                            OLS Regression Results                            
Dep. Variable:        np.log(tip_pct)   R-squared:                       0.235
Model:                            OLS   Adj. R-squared:                  0.232
Method:                 Least Squares   F-statistic:                     74.48
Date:                Wed, 06 Sep 2023   Prob (F-statistic):           8.27e-16
Time:                        17:51:30   Log-Likelihood:                -106.00
No. Observations:                 244   AIC:                             216.0
Df Residuals:                     242   BIC:                             223.0
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     -1.2749      0.056    -22.872      0.0

In [6]:
# Obtenha os valores preditos para log(tip_pct)
tips['pred_log_tip_pct'] = reg_log.fittedvalues

# Convertendo a predição logarítmica de volta para a escala original (exponencial)
tips['pred_tip_pct'] = np.exp(tips['pred_log_tip_pct'])

# Calcule o R^2 na escala original
R2_original_scale = tips[['pred_tip_pct', 'tip_pct']].corr().iloc[0, 1] ** 2
print("R^2 na escala original:", R2_original_scale)

R^2 na escala original: 0.11440988592529767


###  5. ajuste um modelo no log do ```tip_pct``` em função do log de ```net_bill```. Calcule o $R^2$ também em ```tip_pct```.

In [7]:
# Carregando o dataset
tips = sns.load_dataset("tips")
tips['tip_pct'] = tips['tip'] / (tips['total_bill'] - tips['tip'])
tips['net_bill'] = tips['total_bill'] - tips['tip']

# Ajustando o modelo com log-log
reg_log_log = smf.ols('np.log(tip_pct) ~ np.log(net_bill)', data=tips).fit()

# Calculando R^2 na escala original de tip_pct
tips['predicted_log_tip_pct'] = reg_log_log.fittedvalues
tips['predicted_tip_pct'] = np.exp(tips['predicted_log_tip_pct'])
R2_log_log = tips[['tip_pct', 'predicted_tip_pct']].corr().iloc[0, 1]**2

print("R^2 na escala original com log-log:", R2_log_log)

R^2 na escala original com log-log: 0.30894874740322276


###  6. qual modelo resultou no melhor ajuste?

O modelo que resultou no melhor ajuste foi o modelo log-log com log de tip_pct em função do log de net_bill, com um R2 de 0.3089.